src/main/java/de/tum/cit/ase/ares/api/archunit/SecurityRuleExecutor.java

```diff
@@ -0,0 +1,97 @@
+package de.tum.cit.ase.ares.api.archunit;
+
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.core.importer.ImportOption;
+import org.apiguardian.api.API;
+import org.slf4j.LoggerFactory;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Executes the rules specified in the security configuration.
+ */
+@API(status = API.Status.MAINTAINED)
+public class SecurityRuleExecutor {
+
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(SecurityRuleExecutor.class);
+
+    /**
+     * The class names to exclude from the security rules to improve the performance of the ClassFileImport by 2.5 times
+     */
+    private final Set<String> classNamesToExclude = Set.of(
+            "java.lang.Object",
+            "java.lang.String",
+            "java.lang.StringBuffer",
+            "java.lang.NullPointerException",
+            "java.util.TimeZone",
+            "java.util.SimpleTimeZone"
+    );
+    /**
+     * Returns the security rules to be executed according to the configuration file
+     */
+    private final SecurityRuleFactory securityRuleFactory = new SecurityRuleFactory();
+
+    /**
+     * The path to the student submission
+     */
+    private final String studentSubmissionPackage;
+
+    /**
+     * The class file importer
+     */
+    private final ClassFileImporter classFileImporter = new ClassFileImporter();
+
+    /**
+     * @param studentSubmissionPackage The path to the student submission
+     */
+    private SecurityRuleExecutor(String studentSubmissionPackage) {
+        this.studentSubmissionPackage = studentSubmissionPackage;
+    }
+
+    /**
+     * Creates a new instance of the SecurityRuleExecutor
+     *
+     * @param studentSubmissionPackage The path to the student submission
+     * @return A new instance of the SecurityRuleExecutor
+     */
+    public static SecurityRuleExecutor withinPackage(String studentSubmissionPackage) {
+        return new SecurityRuleExecutor(studentSubmissionPackage);
+    }
+
+    /**
+     * Returns the dependencies of the student submission
+     */
+    public Set<String> getDependencies(JavaClasses classes) {
+        return classes.stream()
+                .map(JavaClass::getTransitiveDependenciesFromSelf)
+                .flatMap(Set::stream)
+                .map(d -> d.getTargetClass().getPackageName())
+                .collect(Collectors.toSet());
+    }
+
+
+    /**
+     * Executes the security rules
+     */
+    public void executeSecurityRules() {
+        Set<String> packages = getDependencies(classFileImporter.importPackages(studentSubmissionPackage));
```

Comment:
The selection of the correct packages is something we need to be quite careful about. Students could create their own `org.junit` package and implement their own versions of some classes there, that then will get prefered by the classloader.

Ares1 solved this by having a well-defined set of trusted packages and verifying that these packages are not present in the student submission. Here we would need to go the other way around since a student could just create a `mypackage` package that would then not be analyzed by ArchUnit.

Reply:
ArchUnit works with JavaBytecode and looks at the classes in target/classes and other packages, we can therefore also here check for which packages student have created within the submission by getting their packages, the question would be what happens if they change it and the code doesn't compile on our side (in the test repository) or changes them somehow to not execute the test but pass immediately. I don't know what would happen then. We would have to check this before the exeuction of the tests.

Reply:
I agree thanks for your input!